### PR TITLE
\yii\base\Security updates

### PR DESCRIPTION
**Pull request by tom--**
This PR attempts to implement everything discussed in
- Use PHP 7 random_bytes() in \base\Security #9878 
- Stop using openssl_random_pseudo_bytes() #9879 
-  \base\Security should use PHP's password_hash() by default #9880 

It also adds LibreSSL as a safe alternative to OpenSSL on any platform and, for Windows only on PHP >= 5.4.0, keeps `openssl_random_pseudo_bytes()` after all because it doesn't actually use OpenSSL and goes to CryptGenRandom instead.

I also added type and range checks to `generateRandomString()`. We don't generally do a lot of input checking in Yii methods but in a security API like this I think it's wise. Otherwise invisible app bugs are unmasked by these checks.

Testing the new `generateRandomKey()` isn't simple. I temporarily instrumented `::$_randomSource` so I could see that the different branches were operating correctly and I tested with
-  5.4.45 
-  5.5.30 with and without mcrypt ext
-  5.6.16 with and without LibreSSL 2.3.1
-  7.0.0

But I don't know how to implement those tests permanently.

**I haven't tested on Windows.**

I could add tests for `generateRandomString()`'s type and range checks. But it's difficult for `generateRandomKey()` because in PHP 7, different exceptions are thrown. I haven't found a solution for that.

### CODE REVIEW

**Comment by samdark** [comment]
Looks good to me. @yiisoft/core-developers need additional review.

**Comment by samdark** [comment]
btw., run tests on Windows 10 with PHP 5.4.41, PHP 5.6.9, PHP 7.0.1. All passed.

**Comment by tom--** [comment]
I'm pretty sure it's safe to ignore the second parameter of 

[`string openssl_random_pseudo_bytes ( int $length [, bool &$crypto_strong ] )`](https://secure.php.net/manual/en/function.openssl-random-pseudo-bytes.php) 

given the other tests we do, but I'd welcome any other opinion. Here's the function itself:

https://github.com/php/php-src/blob/master/ext/openssl/openssl.c#L5466-L5518

Can the function return a string *and* set `$crypto_strong` false.

**Comment by tom--** [comment]
Is Scrutinizer right? Or is this alleged bug actually a Scrutinizer bug?

Scrutinizer

File: framework/base/Security.php [hunk 1]
```diff
@@ -77,8 +77,10 @@ class Security extends Component
      * - 'password_hash' - use of PHP `password_hash()` function with PASSWORD_DEFAULT algorithm.
      *   This option is recommended, but it requires PHP version >= 5.5.0
      * - 'crypt' - use PHP `crypt()` function.
+     * @deprecated Since version 2.0.7, @generatePasswordHash() ignores @passwordHashStrategy and
```
Comment (SilverFire) [inline framework/base/Security.php]: I'm not sure, whether `@generatePasswordHash()` is a valid APIDOC reference. Shouldn't it be `[[generatePasswordHash()]]`?

**Comment by SilverFire** [comment]
:+1:

**Comment by samdark** [comment]
> Can the function return a string and set $crypto_strong false?

Nope based on current code. But you never know what would be implemented in the future so it's better to rely on method intended signature and check for it.

File: framework/base/Security.php (hunk 1 above)
Comment (samdark) [inline framework/base/Security.php]: Yes, it should be like @SilverFire suggested.

**Comment by samdark** [comment]
`$matches` thing would work as expected and Scritinizer basically failed but we may introduce nested ifs to make it more readable. That's not really critical though.

File: framework/CHANGELOG.md [hunk 2]
```diff
@@ -90,6 +90,7 @@ Yii Framework 2 Change Log
 - Chg #9953: `TimestampBehavior::getValue()` changed to make value processing consistent with `AttributeBehavior::getValue()` (silverfire)
 - Chg #6419: Added `yii\web\ErrorHandler::displayVars` make list of displayed vars customizable. `$_ENV` and `$_SERVER` are not displayed by default anymore (silverfire)
 - Chg #9528: Traversable objects are now formatted as arrays in `yii\web\XmlResponseFormatter` to support SPL objects and Generators (MaXL-ru)
+- Chg #9878,9879,9880: Make `\base\Security` use `random_bytes()`, LibreSSL, mcrypt, limit OpenSSL to Windows, and to prefer `password_hash()` over `crypt()`
```
Comment (samdark) [inline framework/CHANGELOG.md]: Need your handle at the end of the line i.e. `(tom--)`

File: framework/base/Security.php [hunk 3]
```diff
@@ -507,6 +565,14 @@ public function generateRandomKey($length = 32)
      */
     public function generateRandomString($length = 32)
     {
+        if (!is_integer($length)) {
```
Comment (samdark) [inline framework/base/Security.php]: Better to use `is_int` since `is_integer` is an alias.

File: framework/base/Security.php (hunk 1 above)
Comment (tom--) [inline framework/base/Security.php]: OK. There's another of these somewhere

File: framework/base/Security.php (hunk 3 above)
Comment (tom--) [inline framework/base/Security.php]: OK

**Comment by samdark** [comment]
Merged https://github.com/yiisoft/yii2/commit/1f41a2c9af106e70076b72991139d29f07a60a37. Huge thanks for taking care of it.

**Comment by tom--** [comment]
Glad to be of help.

Although I am a bit nervous because it was a big change and there were few line comments. I'll see if I can poke anyone else into reviewing it.

**Comment by samdark** [comment]
Yeah. The more people the better.

**Comment by githubjeka** [comment]
@samdark somehow marked `@deprecated` in https://github.com/yiisoft/yii2/wiki/Ideas-for-2.1 or is checked by searching afterwards?

**Comment by samdark** [comment]
,

**Comment by githubjeka** [comment]
I asked about:
```
@deprecated Since version 2.0.7, [[generatePasswordHash()]] ignores [[passwordHashStrategy]] and
uses `password_hash()` when available or `crypt()` when not.

 -    public $passwordHashStrategy = 'crypt';
 +    public $passwordHashStrategy;
```

**Comment by samdark** [comment]
It's marked in IDEs and would be removed afterwards.

**Comment by githubjeka** [comment]
Ok.

P.S. @tom-- thanks for your contribute

**Comment by tom--** [comment]
> This pull request is closed, but the `tom--:9878-9879-9880-base/Security-updates` branch has unmerged commits.

How to see what these changes are?

**Comment by samdark** [comment]
That's composer.lock.

**Pull request closed without merging**